### DISCUSSION TITLE
Need clarification about InputCapture barriers

### DISCUSSION THREAD

**Discussion by Sodivad** [flatpak/xdg-desktop-portal#1316] [General]
Currently the documentation says

> For example, consider two zones of size 1920x1080 with offsets 0,0 and 1920,0, respectively. (i.e. a left-right dual-monitor setup). The following pointer barriers are permitted:
> 
>     top edge of left screen: x1=0, y1=0, x2=1919, y1=0
> 
>     bottom edge of left screen: x1=0, y1=1080, x2=1919, y1=1080
> 
>     top edge of right screen: x1=1920, y1=0, x2=3839, y1=0
> 
>     bottom edge of right screen: x1=1920, y1=1080, x2=3839, y1=1080
> 
>     left edge of left screen: x1=0, y1=0, x2=0, y1=1079
> 
>     right edge of right screen: x1=3840, y1=0, x2=3840, y1=1079
> 
This has no info about  the right edge of the left screen  and left edge of the right screen

My first implementation assumed that
- right edge of the left screen  `x1=1919, y1=0, x2=1919, y2=1079`
- left edge of the right screen  `x1=1920, y1=0, x2=1920, y2=1079`

However when testing I found that input-leap requests
> [2024-03-26T09:26:42] DEBUG: Zone at 1920x1080@0,0
> [2024-03-26T09:26:42] DEBUG: Barrier (top) 0 at 0,0-1919,0
> [2024-03-26T09:26:42] DEBUG: Barrier (right) 1 at 1920,0-1920,1079
> [2024-03-26T09:26:42] DEBUG: Barrier (left) 2 at 0,0-0,1079
> [2024-03-26T09:26:42] DEBUG: Barrier (bottom) 3 at 0,1080-1919,1080
> [2024-03-26T09:26:42] DEBUG: Zone at 1920x1080@1920,0
> [2024-03-26T09:26:42] DEBUG: Barrier (top) 4 at 1920,0-3839,0
> [2024-03-26T09:26:42] DEBUG: Barrier (right) 5 at 3840,0-3840,1079
> [2024-03-26T09:26:42] DEBUG: Barrier (left) 6 at 1920,0-1920,1079
> [2024-03-26T09:26:42] DEBUG: Barrier (bottom) 7 at 1920,1080-3839,1080

(disregarding that 0 is an invalid id) it uses `1920,0-1920,1079` for both which means it's not possible to disambiguate both. 

Looking at the examples, I think there is a similar problem for the bottom edge if there was a vertical layout, as it says `y1=y2=1080` when I would expect 1079 which is used for vertical barriers.

@whot @jadahl

**Comment by jadahl** [comment]
What a coordinate means was clarified here: https://github.com/flatpak/xdg-desktop-portal/pull/1136.

> This has no info about the right edge of the left screen and left edge of the right screen

Hrm, I think that a barrier in between isn't allowed. Can't remember why that was limited. Looking at the GNOME code such a barrier would be denied with a "Adjacent to multiple monitor edges" error message.

**Reply by Sodivad** [reply]
Yes, I do the same but it changes the code a bit.

**Comment by whot** [comment]
ftr, 6a29509dfffca4a04215d715cf28b7577ba141fc changed to no longer specify `0-1920` but `0-1919` instead (and that commit has the reasoning too).

InputLeap's barrier code is extremely simple (I wrote it originally :smile:), it just requests any *possible* barrier and ignores the failed ones. That was easier to write than writing a more complex "can I actually have a barrier here" code but it's also too simple for a layout where the monitors don't align perfectly.
The vertical barrier it requests at 1920 is actually not allowed, you must not have a barrier that is not at the far edge of any direction. the 1920 one would go through the middle of the desktop.

**Reply by Sodivad** [reply]
> ftr, https://github.com/flatpak/xdg-desktop-portal/commit/6a29509dfffca4a04215d715cf28b7577ba141fc changed to no longer specify 0-1920 but 0-1919 instead (and that commit has the reasoning too).

Shouldn't the same be done for the vertical position
`bottom edge of left screen: x1=0, y1=1080, x2=1919, y1=1080`, afaiu `y=1080` is not on the screen and should be `1079`

> The vertical barrier it requests at 1920 is actually not allowed, you must not have a barrier that is not at the far edge of any direction. the 1920 one would go through the middle of the desktop.

I am asking in general about the right edges of screens. Currently the example says also
`right edge of right screen: x1=3840, y1=0, x2=3840, y1=1079`, I think 3840 is not on the screen anymore and should be 
` x1=3839, y1=0, x2=3839, y1=1079`,

Or in the case of a single screen what should be the right edge: `x=1920` or x=`1919`. 

I think it's confusing that currently according to the examples different things seem to apply for horizontal/vertical barriers wrt to monitor dimensions

**Reply by Sodivad** [reply]
I think this would also  match the pretty ASCII art in the PR that @jadahl linked and the wording, otherwise the barrier would not be on any screen.

**Reply by whot** [reply]
paging part of this back in: the barriers are on the left/top of the pixel but expand the whole pixel on the other dimension. Zoomed in, a barrier going to 0/1919 would look like this:

```
============
    | 1919 |
----+------+
```
and a barrier at 1920/1079 like this 
```
|    | 1919 ||
+----+------|| 
|    | 1919 || 
+----+------|| 
```

So a horizontal barrier is always inclusive `[x1, x2]` (hence it has to stop at 1919, not 1920) and a vertical barrrier is inclusive `[y1, y2]` (hence it has to stop at 1079, not 1080 - as you said above).

If you put a vertical barrier on `3939` you're putting it on the pixel border between `3938` and `3939` and that should fail, it's not the edge of the screen. In this case you do need to go "over" by one (same for a bottom barrier).

Because we don't allow for overlaps going over is not ambiguous since a 1920 barrier cannot be shared with the right monitor, otherwise it'd fail.

**Reply by Sodivad** [reply]
Ah ok that is where my confusion stemmed from, the barriers only expanding the whole pixel  in their direction but not in the other one. Thanks!

**Comment by whot** [comment]
I think this can be closed based on https://github.com/flatpak/xdg-desktop-portal/discussions/1316#discussioncomment-9151874 (don't have the permissions myself)
